### ISSUE TITLE
Require noreturn to call a noreturn function

### ISSUE THREAD

**Issue by marler8997** [ziglang/zig#17234]
This is a proposal to require the `noreturn` keyword before calling any `noreturn` function.  This makes it explicit at the callsite that the function being called will never return, i.e.

```zig
fn specialCase() noreturn { ... }

if (some_condition) {
    noreturn specialCase();
}

// code that should never run if some_condition was true
```

This elevates "noreturn" control flow to Zig's syntax rather than it being inferred from the function definition.

**Comment by mlugg** [comment]
What's the motivation for this proposal? How does it improve the language?

**Comment by marler8997** [comment]
It makes the "noreturn" control flow explicit at the call site.

In some cases, code that calls a noreturn function depends on that function being noreturn in order for it to be correct.  Currently this dependence is implicit.  Some users may employ a convention to make this more clear such as adding an "exit" suffix to the function name being called, i.e.

```zig
fn specialCaseAndExit() noreturn { ... }

if (some_condition) {
    specialCaseAndExit();
}

// code that should never run if some_condition was true
```

Also note that the example above could be modified to use and `if/else`, but that still doesn't solve the situation for other cases such as calling a noreturn function within an expression or depending on a noreturn function exiting the program altogether.

**Comment by rohlem** [comment]
I agree that required marking of `noreturn` paths in code would help readability.

 ~~I once proposed in https://github.com/ziglang/zig/issues/10420#issuecomment-1004481757 to require an `unreachable` statement after a `noreturn` function call - to me this seems like the same idea without requiring new syntax.
Although if syntactic differentiation can result in clearer error diagnostics or simplifies implementation, then that might be worth it~~.
I guess we don't (and maybe won't ever) have compile errors for statically reachable `unreachable` statements, whereas `noreturn` calling a function that isn't `noreturn` should trivially be a compile error, so that's a definitive improvement.

**Comment by rohlem** [comment]
One scenario we might want to consider: What if a callsite knows that a non-noreturn-declared function will not return?

```zig
fn mayExit() void {}

//option A: wrap
fn willExit() noreturn {
  mayExit();
  unreachable; //signals `noreturn` to the compiler without additional rules to the language
}
fn useA() void {
  if (...) return;
  noreturn willExit();
}

//option B: support casting a function call expression to `noreturn`
fn useB() void {
  if (...) return;
  noreturn @as(noreturn, mayExit()); //makes sense to me, might constrain compiler implementation of this proposal
}

//option C: allow `noreturn` for non-noreturn result types
fn useC() void {
  if (...) return;
  noreturn mayExit();
}
```

- To me the compiler check that the `noreturn` expression's argument _is of type `noreturn`_ is half of this feature's value.
  Therefore I do not think option C (calling any function (wrapping any expression) with `noreturn` expression syntax) should be allowed.
- In my eyes option A is unnecessary boilerplate that worsens locality (although it can be fully implemented in a userspace `meta`-like function).
- I think option B is suitably explicit. It probably grows the language specification a little and constrains the compiler implementation, but feels worth it imo - open for discussion.

**Comment by daurnimator** [comment]
It's common for functions to only *sometimes* not return. 
Given that "maybe noreturn" is not part of a return type; I don't think this issue was a language proposal to add it.

So it's only useful when calling functions that *never* return? That seems like a slim subset, and potentially confusing.

What about functions that `noreturn` for some targets; and other targets "maybe noreturn"? e.g.
```zig
fn foo(X: sometime) u32 {
    if (comptime targetSupportsSomeFeature()) {
        longjmp(X.somebuf);
        unreachable;
    }
    dosomebookkeeping();
    return -1;
}
```

**Comment by matklad** [comment]
Note that `comptime unreachable` covers some of the use-cases here:

```zig
const std = @import("std");

fn will_not_return() noreturn {
    while (true) {}
}

pub fn main() !void {
    if (std.crypto.random.boolean()) {
        will_not_return();
        comptime unreachable;
    } else {
        std.debug.print("Lucky!\n", .{});
    }
}
```


While not _required_, the caller can use it to make it syntactically obvious (and checked at comptime) that callee doesn't return.

**Comment by andrewrk** [comment]
Rejecting for dubious bug-squashing value, and increased language complexity. Programmers who wish to document noreturn callsites are encouraged to use `comptime unreachable`.
